### DISCUSSION TITLE
[Experiments] [2025-01-31] lightGCN 아이템 임베딩 변경

### DISCUSSION THREAD

**Discussion by SayOny** [boostcampaitech7/level4-recsys-finalproject-hackathon-recsys-05-lv3#35] [Experiment]
### 실험 이름

아이템 임베딩 변경

### 작성자

박세연

### 배경 및 목적

기존 lightGCN 모델에 아이템 메타데이터 정보를 포함하고자 했습니다.
제가 생각한 두가지 방법은 다음과 같습니다.

 1. item 임베딩에 메타데이터 임베딩을 합쳐서 학습하는 방법
 먼저, loader 객체를 생성할 때 메타데이터를 가져와 저장합니다. 현재 실험에서는 기본적으로 movies.dat에서 제공되는 장르 정보만 사용했습니다. 따라서 장르 정보를 멀티 핫 인코딩 한 뒤, 아이템 별 장르를 매핑합니다. (딕셔너리)
다음으로, 모델 클래스 객체를 생성할 때 장르 임베딩을 추가하고, 아이템 임베딩과 장르 임베딩을 합한 것을 기존의 latent dimension으로 축소시키는 선형 레이어를 추가합니다.
마지막으로, Propagate 과정에서 
all_emb = torch.cat([users_emb, items_emb], dim=0)  를
all_emb = torch.cat([users_emb, projected_items_emb], dim=0)  
으로 변경합니다.

2. 사용자-아이템 그리고 아이템-장르 관계를 추가하여 그래프를 그리는 방법
아까와 비슷하지만
getSparseGraph라는 함수에서 아이템-장르 관계를 추가하는 방법이다.  

### 실험 설계

데이터셋: 
모델: LightGCN 모델 구조 변형
평가지표: Validation 시 BRP Loss 그리고 Precision, Recall, NDCG@20

### 실험 결과

<img width="924" alt="스크린샷 2025-01-31 09 11 09" src="https://github.com/user-attachments/assets/38e7dd04-8e91-45f4-85d7-891e5193aa8c" />

위 사진에서 볼 수 있듯이 바닐라 모델과 비교하여 test set에 대해서 도입한 두가지 방법의 결과가 무시할 수 없는 정도로 낮아진 것을 확인할 수 있다.

<br>

다만 아래의 그림에서 볼 수 있듯이 Train 시에 Validation Dataset에 대한 BPR Loss가 바닐라 모델보다 작은 것을 확인할 수 있다. 특히 첫번째 방법에서 Loss가 확실히 작다.
<img width="937" alt="스크린샷 2025-01-31 09 12 32" src="https://github.com/user-attachments/assets/81975cc9-2508-4985-81e8-62ef6550ef3c" />

또한 Validation set에 대한 Precision, Recall, NDCG @ 20의 결과가 바닐라 모델보다 높다. 

이를 통해 두가지 방법에 대해 과적합이 일어났거나, 코드 수정이 잘못되어 잘 작동하지 않는 상황이라고 생각이 든다.

### 결론 및 인사이트

LightGCN 모델에 아이템 메타데이터를 추가하기 위한 두 가지 방법 모두 과적합이 일어나 test dataset에서 성능이 낮아지는 것을 확인할 수 있었다. 

다만, 앞으로 데이터셋 전처리 방법이 바뀔 수 있으므로 이 방법들을 구현한 코드를 따로 남겨두고 검토하며, 다른 방법을 모색하는 계획의 가져야겠다.

### 다음 단계

1. 기존 설계한 코드를 다시 한 번 검토
2. 몇가지 선행 연구가 있는 것으로 아는데, 관련해서 찾아보거나 새로운 방법을 찾고 구현하기
